pkg/controller/rds/dbparametergroup/zz_controller.go

```diff
@@ -0,0 +1,236 @@
+}
+
+type option func(*external)
+
+func newExternal(kube client.Client, client svcsdkapi.RDSAPI, opts []option) *external {
+	e := &external{
+		kube:           kube,
+		client:         client,
+		preObserve:     nopPreObserve,
+		postObserve:    nopPostObserve,
+		lateInitialize: nopLateInitialize,
+		isUpToDate:     alwaysUpToDate,
+		filterList:     nopFilterList,
+		preCreate:      nopPreCreate,
+		postCreate:     nopPostCreate,
+		preDelete:      nopPreDelete,
+		postDelete:     nopPostDelete,
+		preUpdate:      nopPreUpdate,
+		postUpdate:     nopPostUpdate,
+	}
+	for _, f := range opts {
+		f(e)
+	}
+	return e
+}
+
+type external struct {
+	kube           client.Client
+	client         svcsdkapi.RDSAPI
+	preObserve     func(context.Context, *svcapitypes.DBParameterGroup, *svcsdk.DescribeDBParameterGroupsInput) error
+	postObserve    func(context.Context, *svcapitypes.DBParameterGroup, *svcsdk.DescribeDBParameterGroupsOutput, managed.ExternalObservation, error) (managed.ExternalObservation, error)
+	filterList     func(*svcapitypes.DBParameterGroup, *svcsdk.DescribeDBParameterGroupsOutput) *svcsdk.DescribeDBParameterGroupsOutput
+	lateInitialize func(*svcapitypes.DBParameterGroupParameters, *svcsdk.DescribeDBParameterGroupsOutput) error
+	isUpToDate     func(*svcapitypes.DBParameterGroup, *svcsdk.DescribeDBParameterGroupsOutput) (bool, error)
+	preCreate      func(context.Context, *svcapitypes.DBParameterGroup, *svcsdk.CreateDBParameterGroupInput) error
+	postCreate     func(context.Context, *svcapitypes.DBParameterGroup, *svcsdk.CreateDBParameterGroupOutput, managed.ExternalCreation, error) (managed.ExternalCreation, error)
+	preDelete      func(context.Context, *svcapitypes.DBParameterGroup, *svcsdk.DeleteDBParameterGroupInput) (bool, error)
+	postDelete     func(context.Context, *svcapitypes.DBParameterGroup, *svcsdk.DeleteDBParameterGroupOutput, error) error
+	preUpdate      func(context.Context, *svcapitypes.DBParameterGroup, *svcsdk.ModifyDBParameterGroupInput) error
+	postUpdate     func(context.Context, *svcapitypes.DBParameterGroup, *svcsdk.DBParameterGroupNameMessage, managed.ExternalUpdate, error) (managed.ExternalUpdate, error)
+}
+
+func nopPreObserve(context.Context, *svcapitypes.DBParameterGroup, *svcsdk.DescribeDBParameterGroupsInput) error {
+	return nil
+}
+func nopPostObserve(_ context.Context, _ *svcapitypes.DBParameterGroup, _ *svcsdk.DescribeDBParameterGroupsOutput, obs managed.ExternalObservation, err error) (managed.ExternalObservation, error) {
+	return obs, err
+}
+func nopFilterList(_ *svcapitypes.DBParameterGroup, list *svcsdk.DescribeDBParameterGroupsOutput) *svcsdk.DescribeDBParameterGroupsOutput {
+	return list
+}
+
+func nopLateInitialize(*svcapitypes.DBParameterGroupParameters, *svcsdk.DescribeDBParameterGroupsOutput) error {
+	return nil
+}
+func alwaysUpToDate(*svcapitypes.DBParameterGroup, *svcsdk.DescribeDBParameterGroupsOutput) (bool, error) {
+	return true, nil
+}
```

Comment:
In order to enable update, you need to create a `isUpToDate` function in `setup.go` whose signature is same as this one. Then by comparing the values in `DescribeDBParameterGroupsOutput` and `DBParameterGroup` objects, the function will return `true` or `false`. If it decides that an update is due, only then an update will be done, i.e. `ModifyDBParametersGroupWithContext` call. Though I see that `DBParameterGroup` CRD does not contain parameters field, so you need to add it to `CustomDBParameterGroup`.

So, the next steps would be:
* Implement `isUpToDate`.
* Add the `Parameters` field to `CustomDBParameterGroup`. You can copy the struct from `ModifyDBParametersGroupInput`.
* In `preUpdate`, set the values of `obj` from the field you added in `CustomDBParameterGroup`.
* Manually test all the functionality with an example YAML.

Reply:
Hi @muvaf,
in preUpdate, when trying to set the values of obj, i get the following error:

Reply:
Yes the structs are different. You need to iterate over the array and add to the target array one by one by constructing an `*rds.Parameter` object per entry.